reset.c

```diff
@@ -0,0 +1,849 @@
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE
+
+*/
+
+#include "samtools.h"
+#include "htslib/sam.h"
+#include "sam_opts.h"
+#include "htslib/thread_pool.h"
+#include "htslib/khash.h"
+
+KHASH_SET_INIT_INT(aux_set);        //set hold the aux tags to be retained
+KHASH_SET_INIT_STR(pg_set);         //set hold the pg tags to be removed
+
+#define TAGNUM(X) (((X)[0] << 8) | (X)[1])  //to create key for aux tags, like type key in htslib
+
+typedef struct conf_data
+{
+    int keepRGs;                    //RG line handling
+    int keepPGs;                    //PG line handling
+    khash_t(aux_set) *auxtagsin;    //SET that holds the aux tags to be retained
+    khash_t(pg_set) *pgout;         //SET that holds the pg tags to be removed
+
+    int tmpfield;                   //tmp field to hold any invocation specific data for the callback
+} conf_data;
+
+/// usage - print the usage
+/** @param fp pointer to the file / terminal to which usage to be dumped
+returns nothing
+*/
+static void usage(FILE *fp)
+{
+    fprintf(fp,
+            "Usage: samtools reset [options]\n\
+      -i <file path>        input file\n\
+      -o <file path>        output file\n\
+      -k <file path>        aux tags to be retained in csv file, optional\n\
+      -r <file path>        aux tags to be rejected in csv file, optional\n\
+      --reject-PG [<file path>] pg names to be rejected in csv file, optional\n\
+      --no-RG               to have RG lines or not, optional\n");
+
+    sam_global_opt_help(fp, "--.--.--");
+    return;
+}
+
+/// getdefaultauxtags - add default aux tags to the collection
+/** @param config pointer to conf_data which holds the aux_set pointer; if the aux_set exists, clear and create new one
+returns 1 on failure and 0 on success
+*/
+int getdefaultauxtags(conf_data *config)
+{
+    int ret = 0;
+
+    if (!config) {
+        return 1;
+    }
+
+    if (config->auxtagsin) {      //free if already exists
+        kh_destroy(aux_set, config->auxtagsin);
+        config->auxtagsin = NULL;
+    }
+
+    config->auxtagsin = kh_init(aux_set);    //initialize the new one
+
+    //add defaults, they may get overridden by reject option
+    kh_put(aux_set, config->auxtagsin, TAGNUM("BC"), &ret);   //barcode
+    kh_put(aux_set, config->auxtagsin, TAGNUM("CB"), &ret);   //cell identifier
+    //TODO add tags which are kept for backward compatibility?
+    kh_put(aux_set, config->auxtagsin, TAGNUM("LB"), &ret);   //library
+    kh_put(aux_set, config->auxtagsin, TAGNUM("MI"), &ret);   //molecular identifier
```

Comment:
These defaults need to be added to the usage statement.  The man page is also too wishy-washy when describing this.  It needs to be explicit stating the default list is exactly these 4 items and not implying there may be other unlisted ones too.

You can see examples of how we handle lists like this in the `samtools mpileup` usage:

```
  --ff, --excl-flags STR|INT  filter flags: skip reads with any of the mask bits set
                                            [UNMAP,SECONDARY,QCFAIL,DUP]
```

Note with `--excl-flags` it defaults to that list of 4 flags, and the user specifying a new list *replaces* the default rather than appending to it. Compare this to `samtools view --keep-tag` which doesn't replace as it appends to the list, but the list starts off empty.

I'm not sure how best to handle this.  There are both pros and cons to replacing vs appending, and there is also the option of doing both where the first usage is an explicit replace, and all subsequent are append.  That way it has a default list, but as soon as you specify any relevant command line options the default is void.  I'm open persuation either way.  Maybe @whitwham or @daviesrob have views on the best CLI for this?